File: src-tauri/src/storage.rs
use crate::types::{AppSettings, ProjectHistoryEntry, WindowBounds};
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::{
  env,
  fs,
  path::{Path, PathBuf},
};

const APP_NAME: &str = "auto-dev-launcher-gui";
const HISTORY_FILE: &str = "project-history.json";
const SETTINGS_FILE: &str = "app-settings.json";

#[derive(Clone)]
pub struct StorageManager {
  storage_path: PathBuf,
}

impl StorageManager {
  pub fn new() -> Result<Self, String> {
    let storage_path = get_storage_path();
    fs::create_dir_all(&storage_path).map_err(|error| format!("鏃犳硶鍒涘缓瀛樺偍鐩綍: {error}"))?;

    Ok(Self { storage_path })
  }

  pub fn load_project_history(&self) -> Result<Vec<ProjectHistoryEntry>, String> {
    Ok(self
      .read_json_with_fallback::<Vec<ProjectHistoryEntry>>(&self.history_file_path())?
      .unwrap_or_default())
  }

  pub fn save_project_history(&self, history: &[ProjectHistoryEntry]) -> Result<(), String> {
    self.atomic_write_json(&self.history_file_path(), history)
  }

  pub fn load_settings(&self) -> Result<AppSettings, String> {
    Ok(self
      .read_json_with_fallback::<AppSettings>(&self.settings_file_path())?
      .unwrap_or_default())
  }

  pub fn save_settings(&self, settings: &AppSettings) -> Result<(), String> {
    self.atomic_write_json(&self.settings_file_path(), settings)
  }

  pub fn save_window_bounds(&self, bounds: &WindowBounds) -> Result<(), String> {
    let mut settings = self.load_settings().unwrap_or_default();
    settings.window_bounds = bounds.clone();
    self.save_settings(&settings)
  }

  fn history_file_path(&self) -> PathBuf {
    self.storage_path.join(HISTORY_FILE)
  }

  fn settings_file_path(&self) -> PathBuf {
    self.storage_path.join(SETTINGS_FILE)
  }

  fn atomic_write_json<T>(&self, file_path: &Path, value: &T) -> Result<(), String>
  where
    T: Serialize + ?Sized,
  {
    let content = serde_json::to_string_pretty(value)
      .map_err(|error| format!("鏃犳硶搴忓垪鍖栧瓨鍌ㄦ暟鎹? {error}"))?;

    let tmp_path = file_path.with_extension("tmp");
    let bak_path = file_path.with_extension("bak");

    fs::write(&tmp_path, content).map_err(|error| format!("鏃犳硶鍐欏叆涓存椂鏂囦欢: {error}"))?;

    if file_path.exists() {
      let _ = fs::copy(file_path, &bak_path);
      let _ = fs::remove_file(file_path);
    }

    fs::rename(&tmp_path, file_path).map_err(|error| format!("鏃犳硶鏇挎崲瀛樺偍鏂囦欢: {error}"))?;

    Ok(())
  }

  fn read_json_with_fallback<T>(&self, file_path: &Path) -> Result<Option<T>, String>
  where
    T: DeserializeOwned + Serialize,
  {
    if file_path.exists() {
      match fs::read_to_string(file_path) {
        Ok(content) => match serde_json::from_str::<T>(&content) {
          Ok(value) => return Ok(Some(value)),
          Err(_) => {}
        },
        Err(_) => {}
      }
    }

    let bak_path = file_path.with_extension("bak");
    if !bak_path.exists() {
      return Ok(None);
    }

    let content = fs::read_to_string(&bak_path)
      .map_err(|error| format!("鏃犳硶璇诲彇澶囦唤鏂囦欢: {error}"))?;
    let value = serde_json::from_str::<T>(&content)
      .map_err(|error| format!("澶囦唤鏂囦欢鏍煎紡鏃犳晥: {error}"))?;

    self.atomic_write_json(file_path, &value)?;

    Ok(Some(value))
  }
}

fn get_storage_path() -> PathBuf {
  if cfg!(target_os = "windows") {
    let base = env::var("APPDATA").unwrap_or_else(|_| {
      let home = env::var("USERPROFILE").unwrap_or_else(|_| ".".to_string());
      format!("{home}\\AppData\\Roaming")
    });

    return PathBuf::from(base).join(APP_NAME);
  }

  if cfg!(target_os = "macos") {
    let home = env::var("HOME").unwrap_or_else(|_| ".".to_string());
    return PathBuf::from(home)
      .join("Library")
      .join("Application Support")
      .join(APP_NAME);
  }

  let home = env::var("HOME").unwrap_or_else(|_| ".".to_string());
  PathBuf::from(home).join(".config").join(APP_NAME)
}

File: internal/storage/storage.go
package storage

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/60liulin60/auto-dev-launcher-gui/refactor/internal/types"
)

const (
	// appName 与现网 Tauri 版一致，保证用户数据路径兼容
	appName      = "auto-dev-launcher-gui"
	historyFile  = "project-history.json"
	settingsFile = "app-settings.json"
)

// Manager 本地 JSON 持久化（历史 + 设置），支持 .bak 回退与原子写
type Manager struct {
	storagePath string
}

// New 创建存储管理器；目录不存在时自动创建
func New() (*Manager, error) {
	path := getStoragePath()
	if err := os.MkdirAll(path, 0o755); err != nil {
		return nil, fmt.Errorf("无法创建存储目录: %w", err)
	}
	return &Manager{storagePath: path}, nil
}

// NewWithPath 测试用：指定存储根目录
func NewWithPath(path string) (*Manager, error) {
	if err := os.MkdirAll(path, 0o755); err != nil {
		return nil, err
	}
	return &Manager{storagePath: path}, nil
}

// StoragePath 返回存储根路径
func (m *Manager) StoragePath() string {
	return m.storagePath
}

// LoadProjectHistory 读取项目历史；损坏时尝试 .bak
func (m *Manager) LoadProjectHistory() ([]types.ProjectHistoryEntry, error) {
	value, err := readJSONWithFallback[[]types.ProjectHistoryEntry](m.historyFilePath())
	if err != nil {
		return nil, err
	}
	if value == nil {
		return []types.ProjectHistoryEntry{}, nil
	}
	return *value, nil
}

// SaveProjectHistory 原子写入项目历史
func (m *Manager) SaveProjectHistory(history []types.ProjectHistoryEntry) error {
	return atomicWriteJSON(m.historyFilePath(), history)
}

// LoadSettings 读取应用设置
func (m *Manager) LoadSettings() (types.AppSettings, error) {
	value, err := readJSONWithFallback[types.AppSettings](m.settingsFilePath())
	if err != nil {
		return types.AppSettings{}, err
	}
	if value == nil {
		return types.DefaultAppSettings(), nil
	}
	return *value, nil
}

// SaveSettings 原子写入应用设置
func (m *Manager) SaveSettings(settings types.AppSettings) error {
	return atomicWriteJSON(m.settingsFilePath(), settings)
}

// SaveWindowBounds 更新并保存窗口边界
func (m *Manager) SaveWindowBounds(bounds types.WindowBounds) error {
	settings, err := m.LoadSettings()
	if err != nil {
		settings = types.DefaultAppSettings()
	}
	settings.WindowBounds = bounds
	return m.SaveSettings(settings)
}

func (m *Manager) historyFilePath() string {
	return filepath.Join(m.storagePath, historyFile)
}

func (m *Manager) settingsFilePath() string {
	return filepath.Join(m.storagePath, settingsFile)
}

// atomicWriteJSON 先写 .tmp，再备份旧文件为 .bak，最后 rename
func atomicWriteJSON(filePath string, value any) error {
	content, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return fmt.Errorf("无法序列化存储数据: %w", err)
	}

	tmpPath := filePath + ".tmp"
	bakPath := filePath + ".bak"

	if err := os.WriteFile(tmpPath, content, 0o644); err != nil {
		return fmt.Errorf("无法写入临时文件: %w", err)
	}

	if _, err := os.Stat(filePath); err == nil {
		_ = copyFile(filePath, bakPath)
		_ = os.Remove(filePath)
	}

	if err := os.Rename(tmpPath, filePath); err != nil {
		return fmt.Errorf("无法替换存储文件: %w", err)
	}
	return nil
}

func readJSONWithFallback[T any](filePath string) (*T, error) {
	if data, err := os.ReadFile(filePath); err == nil {
		var value T
		if err := json.Unmarshal(data, &value); err == nil {
			return &value, nil
		}
	}

	bakPath := filePath + ".bak"
	data, err := os.ReadFile(bakPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("无法读取备份文件: %w", err)
	}

	var value T
	if err := json.Unmarshal(data, &value); err != nil {
		return nil, fmt.Errorf("备份文件格式无效: %w", err)
	}

	// 尝试把备份恢复为主文件
	_ = atomicWriteJSON(filePath, value)
	return &value, nil
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, data, 0o644)
}

// getStoragePath 返回与现网一致的用户数据目录
func getStoragePath() string {
	if base := os.Getenv("APPDATA"); base != "" {
		return filepath.Join(base, appName)
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return filepath.Join(".", appName)
	}
	return filepath.Join(home, "AppData", "Roaming", appName)
}
